dask/layers.py

```diff
@@ -1239,3 +1245,314 @@ def __repr__(self):
+        for key in keys:
+            try:
+                _name, _split = key
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            splits.add(_split)
+        return splits
+
+    def _cull(self, output_partitions):
+        return DataFrameTreeReduction(
+            self.name,
+            self.name_input,
+            self.npartitions_input,
+            self.concat_func,
+            self.tree_node_func,
+            finalize_func=self.finalize_func,
+            split_every=self.split_every,
+            split_out=self.split_out,
+            output_partitions=output_partitions,
+            tree_node_name=self.tree_node_name,
+            annotations=self.annotations,
+        )
+
+    def cull(self, keys, all_keys):
+        """Cull a DataFrameTreeReduction HighLevelGraph layer"""
+        deps = {
+            (self.name, 0): {
+                (self.name_input, i) for i in range(self.npartitions_input)
+            }
+        }
+        output_partitions = self._keys_to_splits(keys)
+        if output_partitions != set(self.output_partitions):
+            culled_layer = self._cull(output_partitions)
+            return culled_layer, deps
+        else:
+            return self, deps
+
+    def __dask_distributed_pack__(self, *args, **kwargs):
+        from distributed.protocol.serialize import to_serialize
+
+        # Pickle the (possibly) user-defined functions here.
+        # We also wrap kwargs into the functions themselves
+        # to simplify serialization
+        _concat_func = to_serialize(self.concat_func)
+        _tree_node_func = to_serialize(self.tree_node_func)
+        if self.finalize_func:
+            _finalize_func = to_serialize(self.finalize_func)
+        else:
+            _finalize_func = None
+
+        return {
+            "name": self.name,
+            "name_input": self.name_input,
+            "npartitions_input": self.npartitions_input,
+            "concat_func": _concat_func,
+            "tree_node_func": _tree_node_func,
+            "finalize_func": _finalize_func,
+            "split_out": self.split_out,
+            "output_partitions": self.output_partitions,
+            "tree_node_name": self.tree_node_name,
+        }
+
+    @classmethod
+    def __dask_distributed_unpack__(cls, state, dsk, dependencies):
+        from distributed.protocol.serialize import to_serialize
+
+        # Materialize the layer
+        raw = cls(**state)._construct_graph()
+
+        # Convert all keys to strings and dump tasks
+        raw = {stringify(k): stringify_collection_keys(v) for k, v in raw.items()}
+        keys = raw.keys() | dsk.keys()
+        deps = {k: keys_in_tasks(keys, [v]) for k, v in raw.items()}
+
+        # Must use `to_serialize` directly to ensure Serialized functions
+        # are deserialized correctly on worker
+        return {"dsk": toolz.valmap(to_serialize, raw), "deps": deps}
```

Comment:
I'm confused—are we double-`to_serialize`ing the functions? Because when we pack them, they're wrapped in `to_serialize`. In the unpack on the scheduler, are they `Serialized` objects, or actually unpickled functions?

Or is this `to_serialize`-ing the entire task tuple, including the nested `Serialized` object?

A comment here might help, especially when we get to refactor all this with https://github.com/dask/distributed/issues/5581.

Reply:
>In the unpack on the scheduler, are they Serialized objects, or actually unpickled functions?

They will be `Serialized` objects.

We are marking the entire task tuple as `Serialize`. We cannot use the typical `dumps_task`, because the already-`Serialized` function objects within each task would become "double pickled".  I was hoping the comment above was explaining why we need `to_serialize`, but I agree that it's not great. Do you have a nicer explanation in mind?   Maybe I can just say that "we need the serialization pass to avoid double-pickling the already `Serialized` function objects in the task"?

Reply:
Ok, how about we have the comment say that "because the task-tuples contain `Serialized` objects instead of real functions, `dumps_task` would go down the path of wrapping the entire tuple in `to_serialize`. So we just save it the trouble. When it arrives on a worker, both the `Serialized` task-tuples and the `Serialized` functions nested within them _should_ be deserialzed automatically by the comm."

Reply:
Okay - I'm updating the comment to be very close to your suggestion.  However, I should clarify that the code was not working when I used `dumps_task`.  The primary task function will not be a `Serialized` object, it will be and explicit `operator.getitem` or `toolz.pipe`.  Therefore, `dumps_task` will not use `to_serialize` here.

Reply:
Ohh okay interesting. So the problem is that `dumps_task` _would_ think it's a task (because there's a `getitem` or `pipe` as the first callable), but all the arguments are `Serialized`. That makes way more sense.

Reply:
Comment looks great!